src/com/amazon/corretto/crypto/provider/AesCfbSpi.java

```diff
@@ -0,0 +1,322 @@
+  }
+
+  @Override
+  protected void engineSetMode(String mode) throws NoSuchAlgorithmException {
+    if (!"CFB".equalsIgnoreCase(mode)) {
+      throw new NoSuchAlgorithmException();
+    }
+  }
+
+  @Override
+  protected void engineSetPadding(String padding) throws NoSuchPaddingException {
+    if (!"NoPadding".equalsIgnoreCase(padding)) {
+      throw new NoSuchPaddingException();
+    }
+  }
+
+  @Override
+  protected int engineGetBlockSize() {
+    return BLOCK_SIZE_IN_BYTES;
+  }
+
+  @Override
+  protected int engineGetOutputSize(final int inputLen) {
+    return inputLen; // No padding, so output len is always equal to inputLen
+  }
+
+  @Override
+  protected byte[] engineGetIV() {
+    return iv.clone();
+  }
+
+  @Override
+  protected AlgorithmParameters engineGetParameters() {
+    return null;
+  }
+
+  @Override
+  protected void engineInit(final int opmode, final Key key, final SecureRandom random)
+      throws InvalidKeyException {
+    try {
+      byte[] newIv = new byte[IV_SIZE_IN_BYTES];
+      random.nextBytes(newIv);
+      engineInit(opmode, key, new IvParameterSpec(newIv), random);
+    } catch (InvalidAlgorithmParameterException e) {
+      throw new InvalidKeyException("Failed to initialize with random IV", e);
+    }
+  }
+
+  @Override
+  protected void engineInit(
+      final int opmode,
+      final Key key,
+      final AlgorithmParameterSpec params,
+      final SecureRandom random)
+      throws InvalidKeyException, InvalidAlgorithmParameterException {
+    if (!(params instanceof IvParameterSpec)) {
+      throw new InvalidAlgorithmParameterException("Params must be an instance of IvParameterSpec");
+    }
+    final IvParameterSpec ivParameterSpec = (IvParameterSpec) params;
+    final byte[] ivBytes = ivParameterSpec.getIV();
+    if (ivBytes.length != IV_SIZE_IN_BYTES) {
+      throw new InvalidAlgorithmParameterException(
+          "Provided IV must be of length " + IV_SIZE_IN_BYTES);
+    }
+
+    if (!"RAW".equalsIgnoreCase(key.getFormat())) {
+      throw new InvalidKeyException("Key's format must be RAW");
+    }
+    final byte[] keyBytes = key.getEncoded();
+    if (keyBytes == null) {
+      throw new InvalidKeyException("Key must support encoding");
+    }
+    if (keyBytes.length != KEY_LEN_AES128 && keyBytes.length != KEY_LEN_AES256) {
+      throw new InvalidKeyException(
+          "Key length must be " + KEY_LEN_AES128 + " or " + KEY_LEN_AES256);
+    }
+
+    init(opmode, keyBytes, ivBytes);
+  }
+
+  @Override
+  protected void engineInit(
+      final int opmode, final Key key, AlgorithmParameters params, final SecureRandom random)
+      throws InvalidKeyException, InvalidAlgorithmParameterException {
+    throw new UnsupportedOperationException(
+        "IV must be provided by passing an instance of IvParameterSpec.");
+  }
+
+  private void init(final int opmode, final byte[] keyBytes, final byte[] ivBytes) {
+    this.opMode = (opmode == Cipher.ENCRYPT_MODE) ? ENC_MODE : DEC_MODE;
```

Comment:
This allows the user to specify other modes like Wrap/Unwrap mode and end up with Decrypt. I'd rather we check for decrypt explicitly and throw an invalid parameter where the caller passes in junk.